docs/DesignDocs/shared_memory.md

```diff
@@ -0,0 +1,192 @@
+The Open Enclave shared memory model
+====================================
+
+This document proposes a *shared memory model* that will improve
+interoperability across different secure hardware platforms, initially
+including SGX and TrustZone.
+
+Motivation
+----------
+
+Ideally, applications developed with the Open Enclave API should run on all
+supported platforms. But unfortunately, the platforms themselves may have
+incompatible shared memory models. For example, SGX implicit shares all
+untrusted memory with the enclave, whereas TrustZone shares untrusted memory
+explicitly through programmatic measures (by default, no untrusted memory
+is shared).
+
+This problem is partly addressed by the PR entitled [Enforcing full
+serialization in EDL](https://github.com/openenclave/openenclave/pull/2176),
+which proposes extending **edger8r** to print warnings for EDL specifications
+with parameters that cannot be fully serialized. Such specifications produce
+applications that implicitly share untrusted memory with the enclave. Aside
+from introducing potential security flaws, these applications are not portable
+to TrustZone and any future platforms that employ a similar model. By forcing
+full serialization, the need for shared memory can be mostly eliminated. But
+some applications need to pass untrusted shared memory to enclaves. Examples
+include:
+
+- Ring buffers
+- Context-switchless calls
+- Efficient transfer of large memory objects
+
+Whatever the reasons, a mechanism is needed for sharing untrusted memory with
+the enclave and that mechanism must be interoperable across secure hardware
+platforms. Further, the mechanism must be explicit, so operations are needed
+for managing untrusted memory.
+
+User Experience
+---------------
+
+The host may share untrusted memory with the enclave through the use of the
+**user_check** EDL attribute. For example:
+
+```c++
+enclave
+{
+    trusted
+    {
+        void share_buffer_ecall(
+            [user_check] void* untrusted_memory,
+            size_t untrusted_memory_size);
+    };
+};
+```
+
+We assume in this example that the enclave needs to read or write the
+**untrusted_memory** buffer. Prior to this proposal, SGX applications
+could pass any pointer to untrusted memory. For example:
+
+```c++
+    static unsigned char buf[4096];
+
+    share_buffer_ecall(buf, sizeof(buf));
+```
+
+But this application is not portable to TrustZone since **buf** is not visible
+within the enclave. To make this work on TrustZone, the buffer should be
+explicitly allocated from a shared memory pool as follows.
+
+```c++
+    const size_t n = 4096;
+    unsigned char* buf;
+
+    if (oe_allocate_shared_memory(enclave, n, &buf) != OE_RESULT_OK)
```

Comment:
Is buf guaranteed to be initialized if failure is returned?  Since there's no return; in line 76, the sample code will fall through into the memset in line 79 causing a crash if not.   You probably want a return; after line 76.